### PR TITLE
Add mocha tests linting

### PR DESCRIPTION
**Pull request by roman-mazur**
As suggested by @Page- 
https://www.flowdock.com/app/rulemotion/resin-tech/messages/5050379

Change-type: minor
Signed-off-by: Roman Mazur <roman@balena.io>

Examples:
```
$ resin-lint --tests ./test/
  ✓ test/00-init.coffee
  ✓ test/01-constants.spec.coffee
  ✓ test/02-db.spec.coffee
  ✓ test/03-config.spec.coffee

✓ Ok! » 0 errors and 0 warnings in 25 files

Mocha tests check FAILED!
File test/01-constants.spec.coffee, line 4: found describe.only

$ resin-lint --typescript --tests ./test/


Mocha tests check FAILED!
File test/local-mode.ts, line 301: found describe.only
File test/local-mode.ts, line 312: found it.only
```

I can check eslint plugin alternatives later - for now this change can help with keeping checks consistent across the repos.

### CODE REVIEW

File: lib/mocha-tests-lint.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+const checkPattern = /((describe)|(it))\.only/;
+
+// Check if files in the test directory don't contain describe.only or it.only statements
+// to ensure that an accidental commit does not prevent all the tests from running.
+export function lintMochaTests(scripts: string[]): MochaListResult {
+	let errorsFound = false;
+	let message = '';
+
+	for (const scriptPath of scripts) {
+		const content = fs.readFileSync(scriptPath, { encoding: 'utf8' });
```
Comment (CameronDiver) [inline lib/mocha-tests-lint.ts]: I'd rather this wasn't a synchronous file access.

Comment (roman-mazur) [reply lib/mocha-tests-lint.ts]: Would you elaborate on that? What potentials problems are thinking about?

Comment (Page-) [reply lib/mocha-tests-lint.ts]: In general it blocks the event loop which means nothing else can happen at the same time, and in this case that'll mean it takes longer to lint compared to if the file reads were done in parallel (potentially significantly longer with lots of mocha files)

File: package.json [hunk 2]
```diff
@@ -31,7 +31,7 @@
     "@types/bluebird": "^3.5.26",
     "@types/depcheck": "^0.6.0",
     "@types/glob": "^5.0.35",
-    "@types/node": "^8.10.45",
+    "@types/node": "^10.14.10",
```
Comment (Page-) [inline package.json]: Why was this necessary?  Mainly asking because it's potentially a breaking change

File: lib/mocha-tests-lint.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// Mocha tests linter results.
+// When isError is set to true, the message contains details about found errors.
+export class MochaListResult {
+	constructor(
+		public readonly message: string,
+		public readonly isError: boolean,
+	) {}
+}
+
+const checkPattern = /((describe)|(it))\.only/;
```
Comment (Page-) [inline lib/mocha-tests-lint.ts]:
```suggestion
const checkPattern = /(describe|it)\.only/;
```

File: package.json (hunk 2 above)
Comment (roman-mazur) [reply package.json]: should be possible to revert - I needed this because the original implementation was using `fs.readDir` variation that was returning `Dirent[]`, and it was introduced in later versions.

But since I rewrote the checker to fit the rest of the lint code, it's not longer necessary.

File: lib/mocha-tests-lint.ts (hunk 1 above)
Comment (roman-mazur) [reply lib/mocha-tests-lint.ts]: Parallelized.

**Comment by roman-mazur** [comment]
@Page- @CameronDiver 
PTAL

**Comment by roman-mazur** [comment]
@balena-ci retest

**Comment by roman-mazur** [comment]
My concern is also that it will not be enough to bump the linter version to get mocha tests check running - linter needs to understand if running the checks makes sense (--tests cmd line flag).

Do you guys think we should implicitly set this flag if the input directory is `"test"`?

File: lib/mocha-tests-lint.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+// Check if files in the test directory don't contain describe.only or it.only statements
+// to ensure that an accidental commit does not prevent all the tests from running.
+export async function lintMochaTests(
+	scripts: string[],
+): Promise<MochaListResult> {
+	let errorsFound = false;
+	let message = '';
+
+	const readFile = Bluebird.Promise.promisify(fs.readFile);
```
Comment (Page-) [inline lib/mocha-tests-lint.ts]:
```suggestion
	const readFile = Bluebird.promisify(fs.readFile);
```

File: lib/mocha-tests-lint.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+	const allChecks = [];
+	for (const scriptPath of scripts) {
+		allChecks.push(
+			readFile(scriptPath)
+				.then(buffer => buffer.toString())
```
Comment (Page-) [inline lib/mocha-tests-lint.ts]:
```suggestion
				readFile(scriptPath, 'utf8')
```

File: lib/mocha-tests-lint.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+export async function lintMochaTests(
+	scripts: string[],
+): Promise<MochaListResult> {
+	let errorsFound = false;
+	let message = '';
+
+	const readFile = Bluebird.Promise.promisify(fs.readFile);
+
+	const allChecks = [];
+	for (const scriptPath of scripts) {
```
Comment (Page-) [inline lib/mocha-tests-lint.ts]: You can do this as
```ts
const allChecks = scripts.map(async (script) => {
	content = await readFile(scriptPath, 'utf8')
	...
})

File: lib/resin-lint.ts [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import * as Promise from 'bluebird';
+import * as BluePromise from 'bluebird';
```
Comment (Page-) [inline lib/resin-lint.ts]:
```suggestion
import * as Bluebird from 'bluebird';
```

File: lib/mocha-tests-lint.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+							errorsFound = true;
+							message += `File ${scriptPath}, line ${ln}: found ${res[0]}\n`;
+						}
+					}
+				}),
+		);
+	}
+
+	try {
+		await Promise.all(allChecks);
```
Comment (CameronDiver) [inline lib/mocha-tests-lint.ts]: `await Bluebird.map` with an async callback would make more sense here.

File: lib/resin-lint.ts (hunk 7 above)
Comment (CameronDiver) [inline lib/resin-lint.ts]: We generally do `import * as Bluebird from 'bluebird';` and you've also done this in mocha-test-lint.ts, so it'd be better to stay consistent.

File: lib/mocha-tests-lint.ts (hunk 6 above)
Comment (roman-mazur) [reply lib/mocha-tests-lint.ts]: Speaking about `readFile(scriptPath, 'utf8')`, it does not compile. Type of `Bluebird.promisify(fs.readFile)` allows only one argument.

File: lib/mocha-tests-lint.ts (hunk 8 above)
Comment (roman-mazur) [reply lib/mocha-tests-lint.ts]: I've used Page's suggestion above.

**Comment by roman-mazur** [comment]
@balena-ci retest

I fixed Bluebird imports and removed `for` loop in the mocha checks implementation.

What do you think now, @Page- @CameronDiver ?

**Comment by roman-mazur** [comment]
Also, any thoughts on doing tests check using the source directory name?
https://github.com/balena-io/node-resin-lint/pull/39#issuecomment-508122008

**Comment by CameronDiver** [comment]
My concern is that I'd only want this to run on commit/prepare. I'm fine with `test(s?)` being used as the directory we search, but I'd like a flag either enabling or disabling this checking. I'm not too familiar with optimist but it looks like you have a flag for these checks, is it automatically enabled? How would I then disable it?

**Comment by roman-mazur** [comment]
@CameronDiver I illustrated usage in supervisor (as you wanted, as a pre-commit hook only) here:
https://github.com/balena-io/balena-supervisor/pull/1031

Does it answer your question?

File: lib/mocha-tests-lint.ts (hunk 6 above)
Comment (Page-) [reply lib/mocha-tests-lint.ts]: Yeah, that's a typing issue because if you did
```ts
const readFile = Bluebird.Promise.promisify(fs.readFile) as any;
```
then it would work fine at runtime.  I normally use `fs.promises.readFile` in node 10+ to get around this because it has the built in types that work nicely but we can't guarantee that for resin-lint :(
